### PR TITLE
Introduce ResilienceStrategyBuilder

### PR DESCRIPTION
**Pull request by martintmk**
### The issue or feature being addressed

This PR adds a `ResilienceStrategyBuilder` class from v8-poc branch. It's used to build the instance of `IResilienceStrategy`.

### Details on the issue fix or feature implementation

The following primitives are added:

- `ResilienceStrategyBuilder`: the builder itself.
- `ResilienceStrategyBuilderContext`: context that caries details for the strategy factory.
- `ResilienceStrategyBuilderOptions`: the options of the builder. These will be epxanded later with more members such as clock, logging, telemetry. 
- `ResilienceStrategyOptions`: the options attached to individual strategy.

I'll add tests once we agree on the API surface.

### Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have successfully run a local build

### CODE REVIEW

File: src/Polly.Core/Builder/IResilienceStrategyBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+namespace Polly.Builder;
+
+/// <summary>
+/// A builder that is used to create a <see cref="IResilienceStrategy"/>.
```
Comment (martincostello) [inline src/Polly.Core/Builder/IResilienceStrategyBuilder.cs:4]:
```suggestion
/// A builder that is used to create an <see cref="IResilienceStrategy"/>.
```

File: src/Polly.Core/Builder/IResilienceStrategyBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+/// </summary>
+/// <remarks>
+/// The builder supports chaining multiple strategies into pipeline of strategies.
```
Comment (martincostello) [inline src/Polly.Core/Builder/IResilienceStrategyBuilder.cs:7]:
```suggestion
/// The builder supports chaining multiple strategies into a pipeline of strategies.
```

File: src/Polly.Core/Builder/ResilienceStrategyBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    /// <inheritdoc/>
+    public IResilienceStrategyBuilder AddStrategy(Func<ResilienceStrategyBuilderContext, IResilienceStrategy> factory, ResilienceStrategyOptions? options = null)
+    {
+        Guard.NotNull(factory);
+
+        if (options != null)
+        {
+            Validator.ValidateObject(options, new ValidationContext(options), true);
```
Comment (martincostello) [inline src/Polly.Core/Builder/ResilienceStrategyBuilder.cs:25]: Maybe name the `true` parameter here to make its purpose more obvious?

File: src/Polly.Core/Builder/ResilienceStrategyBuilderExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+namespace Polly.Builder;
+
+/// <summary>
+/// The extensions for the <see cref="IResilienceStrategyBuilder"/>.
+/// </summary>
+public static class ResilienceStrategyBuilderExtensions
+{
+    /// <summary>
+    /// Adds a already created strategy instance to the builder.
```
Comment (martincostello) [inline src/Polly.Core/Builder/ResilienceStrategyBuilderExtensions.cs:9]:
```suggestion
    /// Adds an already created strategy instance to the builder.
```

File: src/Polly.Core/Builder/ResilienceStrategyOptions.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace Polly.Builder;
+
+/// <summary>
+/// The options associated with the <see cref="IResilienceStrategy"/>.
+/// </summary>
+public class ResilienceStrategyOptions
+{
+    /// <summary>
+    /// Gets or sets the name of the strategy.
```
Comment (juraj-blazek) [inline src/Polly.Core/Builder/ResilienceStrategyOptions.cs:11]: Let's also add the intended usage of these two properties (metering and logging).

File: src/Polly.Core/Utils/ValidationHelper.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+internal static class ValidationHelper
+{
+    public static void ValidateObject(object instance, string mainMessage)
+    {
+        var errors = new List<ValidationResult>();
+
+        if (!Validator.TryValidateObject(instance, new ValidationContext(instance), errors))
+        {
+            StringBuilder stringBuilder = new StringBuilder(mainMessage);
```
Comment (martincostello) [inline src/Polly.Core/Utils/ValidationHelper.cs:15]:
```suggestion
            var stringBuilder = new StringBuilder(mainMessage);
```

**Pull request merged**